shared-bindings/_bit_transpose/__init__.c

```diff
@@ -0,0 +1,100 @@
+/*
+ * This file is part of the Circuit Python project, https://github.com/adafruit/circuitpython
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2018 Roy Hooper
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/obj.h"
+#include "py/runtime.h"
+
+#include "shared-bindings/_bit_transpose/__init__.h"
+
+//| """A fast bit transposition function for parallel NeoPixel strips
+//|
+//| When driving multiple NeoPixel strips from a shift register, the bits
+//| must be re-ordered in a specific way.  This module offers a low-level
+//| routine for performing the transformation."""
+//|
+
+//| def bit_transpose(input: _typing.ReadableBuffer, *, num_strands:int = 8, output: Optional[_typing.WriteableBuffer]=None) -> WriteableBuffer:
+//|     """Convert a sequence of pixel values into a single stream of bytes suitable for sending via a parallel conversion method (PioPixl8)
+//|
+//|     The number of bytes in the input buffer must be a multiple of the number of strands.
+//|
+//|     If specified output buffer must be big enough for the output, ``len(input) * 8  // num_strands``.  To get a properly sized buffer,
+//|     you can also pass in ``output=None`` once, then re-use the return value in future calls.
+//|
+//|     Returns the output buffer if specified (which must be big enough to hold the result), otherwise a freshly allocated buffer."""
```

Comment:
I'd require the output buffer. I generally don't like APIs that allocate themselves because people use it without realizing the impact of it.

Reply:
OK